### PR TITLE
Fix incremental parsing of chunked quoted extensions

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
Before this change, incremental parsing of quoted chunked extensions
was broken for two reasons:

* Http::One::Parser::skipLineTerminator() unexpectedly threw after
  partially received quoted chunk extension value.

* When Http::One::Tokenizer was unable to parse a quoted extension,
  it incorrectly restored the input buffer to the beginning of the
  extension value (instead of the extension itself), thus making
  further incremental parsing iterations impossible.

IMO, the reason for this problem was that Http::One::Tokenizer::qdText()
could not distinguish two cases (returning false in both):

* the end of the quoted string not yet reached

* an input error, e.g., wrong/unexpected character

A possible approach could be to improve Http::One::Tokenizer, making it
aware about "needs more data" state.  However, to be acceptable,
these improvements should be done in the base Parser::Tokenizer
class instead. These changes seem to be non-trivial and could be
done separately and later.

Another approach, used here, is to simplify the complex and error-prone
chunked extensions parsing algorithm, fixing incremental parsing bugs
and still parse incrementally in almost all cases. The performance
regression could be expected only in relatively rare cases of partially
received or malformed extensions.

Also:
* fixed parsing of partial use-original-body extension values
* do not treat an invalid use-original-body as an unknown extension
* optimization: parse use-original-body extension only in ICAP context
  (i.e., where it is expected)
* improvement: added a new API to TeChunkedParser to specify known
  chunked extensions list

### CODE REVIEW

**Comment by eduard-bagdasaryan** [comment]
The related git-dev discussion is [here](http://lists.squid-cache.org/pipermail/squid-dev/2018-October/009475.html).

File: src/http/one/TeChunkedParser.cc [hunk 1]
```diff
@@ -115,55 +116,61 @@ Http::One::TeChunkedParser::parseChunkSize(Http1::Tokenizer &tok)
...
-        return true;
-    }
+        // unknown might have a value token or quoted-string
+        if (quotedStringOrToken(tok, value) && !tok.atEnd())
+            return true;
 
-    return false;
+        return false;
+    }
+    return true;
```
Comment (rousskov) [inline src/http/one/TeChunkedParser.cc]: This `true` return value or the condition preceding it does not look right. For example, If the remaining input contains nothing or just BWS, then the return value should be `false`, not `true` -- we have not parsed the current extension yet, and the rest of that current extension (i.e., extension value) may still be coming.

In other words, the extension "=value" part is indeed optional, but we still need to wait for it if it might get it during the next I/Os.

File: src/http/one/Tokenizer.cc [hunk 2]
```diff
@@ -9,32 +9,14 @@
...
-    if (!skip('"'))
-        return prefix(returnedToken, CharacterSet::TCHAR);
-
-    return qdText(returnedToken, http1p0);
-}
-
-bool
-Http::One::Tokenizer::qdText(SBuf &returnedToken, const bool http1p0)
-{
     // the initial DQUOTE has been skipped by the caller
```
Comment (rousskov) [inline src/http/one/Tokenizer.cc]: A stale comment.

File: src/http/one/Tokenizer.cc [hunk 3]
```diff
@@ -61,12 +43,14 @@ Http::One::Tokenizer::qdText(SBuf &returnedToken, const bool http1p0)
...
+    while (!tok.atEnd()) {
+        if (tok.skip('"'))
+            return true;
 
-        // HTTP/1.1 allows quoted-pair, HTTP/1.0 does not
-        if (!http1p0 && skip('\\')) {
+        SBuf qdText;
+        if (tok.prefix(qdText, tokenChars))
+            returnedToken.append(qdText);
+        else if (!http1p0 && tok.skip('\\')) { // HTTP/1.1 allows quoted-pair, HTTP/1.0 does not
```
Comment (rousskov) [inline src/http/one/Tokenizer.cc]: Remove `else` to simplify the overall logic -- we can handle the quoted pair after handling the plain token prefix above AFAICT, without the need for another loop iteration.

File: src/http/one/Tokenizer.cc [hunk 4]
```diff
@@ -61,12 +43,14 @@ Http::One::Tokenizer::qdText(SBuf &returnedToken, const bool http1p0)
     // best we can do is a conditional reference since http1p0 value may change per-client
     const CharacterSet &tokenChars = (http1p0 ? qdtext1p0 : qdtext1p1);
 
-    for (;;) {
-        SBuf::size_type prefixLen = buf().findFirstNotOf(tokenChars);
-        returnedToken.append(consume(prefixLen));
+    while (!tok.atEnd()) {
+        if (tok.skip('"'))
```
Comment (rousskov) [inline src/http/one/Tokenizer.cc]: Move this `if` statement to the end of the loop to simplify the overall logic (with quoted string _termination_ condition checked _last_) and to optimize the handling of a common case where we can parse the whole quoted string in one loop iteration.

File: src/http/one/Tokenizer.cc [hunk 5]
```diff
@@ -78,32 +62,17 @@ Http::One::Tokenizer::qdText(SBuf &returnedToken, const bool http1p0)
...
-            break; // done
-
-        } else if (atEnd()) {
-            // need more data
-            returnedToken.clear();
-            restoreLastCheckpoint();
-            return false;
+            returnedToken.append(escaped);
+        } else {
+            debugs(24, 8, "invalid bytes for set " << tokenChars.name);
```
Comment (rousskov) [inline src/http/one/Tokenizer.cc]: Should not this throw?

File: src/http/one/Tokenizer.cc [hunk 6]
```diff
@@ -78,32 +62,17 @@ Http::One::Tokenizer::qdText(SBuf &returnedToken, const bool http1p0)
...
             static const CharacterSet qPairChars = CharacterSet::HTAB + CharacterSet::SP + CharacterSet::VCHAR + CharacterSet::OBSTEXT;
             SBuf escaped;
-            if (!prefix(escaped, qPairChars, 1)) {
-                returnedToken.clear();
-                restoreLastCheckpoint();
-                return false;
-            }
-            returnedToken.append(escaped);
-            continue;
+            if (!tok.prefix(escaped, qPairChars, 1))
```
Comment (rousskov) [inline src/http/one/Tokenizer.cc]: This seems inconsistent or even wrong: If the next symbol is invalid, then we should not (silently) return false as if we are simply waiting for more data. We probably should throw an error. 

N.B. It does not matter whether the old code was broken here as well -- we are changing too much of this code to use that as an excuse IMO.

File: src/http/one/Tokenizer.h [hunk 7]
```diff
@@ -9,68 +9,46 @@
...
+class Tokenizer;
+}
 
 namespace Http {
 namespace One {
 
 /**
- * Lexical processor extended to tokenize HTTP/1.x syntax.
+ * Attempt to parse a (token / quoted-string ) lexical construct.
+ *
```
Comment (rousskov) [inline src/http/one/Tokenizer.h]: Please document the return value meaning. For example:
```suggestion
 * \returns true (and sets the value) after successfully parsing the token prefix or the entire quoted string
 * \returns false (and leaves the value intact) if more data is needed to parse the token/string
 * \throws on syntax violations
```

File: src/http/one/Tokenizer.h [hunk 8]
```diff
@@ -9,68 +9,46 @@
...
+namespace Parser {
+class Tokenizer;
+}
 
 namespace Http {
 namespace One {
 
 /**
- * Lexical processor extended to tokenize HTTP/1.x syntax.
+ * Attempt to parse a (token / quoted-string ) lexical construct.
```
Comment (rousskov) [inline src/http/one/Tokenizer.h]: I think we should simplify and detail this, especially the important aspect of the token prefix being indistinguishable from the complete token (at this function scope). For example:
```suggestion
 * Extracts either an HTTP/1 token (prefix) or a complete HTTP/1 quoted-string.
 * Unescapes escaped characters in HTTP/1.1 quoted strings.
 * The caller can (and must) distinguish token prefixes from complete tokens. This
 * is usually done by checking for the trailing garbage and/or expression terminators.
```

File: src/http/one/Tokenizer.cc [hunk 9]
```diff
@@ -78,32 +62,17 @@ Http::One::Tokenizer::qdText(SBuf &returnedToken, const bool http1p0)
...
-        // else, we have an error
-        debugs(24, 8, "invalid bytes for set " << tokenChars.name);
-        returnedToken.clear();
-        restoreLastCheckpoint();
-        return false;
     }
 
-    // found the whole string
-    return true;
+    returnedToken.clear();
```
Comment (rousskov) [inline src/http/one/Tokenizer.cc]: Please avoid touching/modifying the function parameter when returning false.

File: src/http/one/Tokenizer.h [hunk 10]
```diff
@@ -9,68 +9,46 @@
...
-    /// parse the internal component of a quote-string, and terminal DQUOTE
-    bool qdText(SBuf &value, const bool http1p0);
-
-    void checkpoint() { savedCheckpoint_ = buf(); savedStats_ = parsedSize(); }
-    void restoreLastCheckpoint() { undoParse(savedCheckpoint_, savedStats_); }
-
-    SBuf savedCheckpoint_;
-    SBuf::size_type savedStats_;
-};
+bool quotedStringOrToken(::Parser::Tokenizer &tok, SBuf &value, const bool http1p0 = false);
```
Comment (rousskov) [inline src/http/one/Tokenizer.h]: I would s/quotedStringOrToken/tokenOrQuotedString/ to avoid ambiguity.

File: src/http/one/Parser.cc [hunk 11]
```diff
@@ -102,7 +112,7 @@ LineCharacters()
 void
 Http::One::Parser::cleanMimePrefix()
 {
-    Http1::Tokenizer tok(mimeHeaderBlock_);
+    ::Parser::Tokenizer tok(mimeHeaderBlock_);
```
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http/one/Parser.cc]: Done.

**Comment by rousskov** [comment]
I do not see the code changes that correspond to your `Done` comments. Perhaps you forgot to push your changes?

File: src/http/one/Parser.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/http/one/Parser.cc]: This change may need more work. The (implemented) second approach was about removing the namespace prefix in most cases, but I still see `Parser::` prefix inside the `Parser` class. Can that be removed? As a general rule of thumb, we should not say `Foo::bar` inside `Foo`.

File: src/http/one/Parser.h [hunk 12]
```diff
@@ -14,6 +14,10 @@
 #include "http/StatusCode.h"
 #include "sbuf/SBuf.h"
 
+namespace Parser {
+class Tokenizer;
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/http/one/Parser.h]: I still see The same construct in the [PR](https://github.com/squid-cache/squid/pull/310/files#diff-22626d07a8de5e584f9a19ef7c693a4eR15). Please check that you did not miss any occurrences.

File: src/http/one/TeChunkedParser.cc [hunk 13]
```diff
@@ -115,55 +116,61 @@ Http::One::TeChunkedParser::parseChunkSize(Http1::Tokenizer &tok)
...
+            continue; // got one extension; there may be more
+        }
+
+        static const CharacterSet NonLF = (CharacterSet::LF).complement().rename("non-LF");
+        if (tok.skipAll(NonLF) && tok.skip('\n'))
+            throw TexcHere("cannot parse chunk extension"); // <garbage> CR*LF
+
+        return false; // need more data to finish parsing extension
+    }
+    return false; // need data to start parsing extension or CRLF
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/http/one/TeChunkedParser.cc]: AFAICT, you used `the` instead of the requested `an` article. IMO, `an` is more appropriate in this context because, unlike the other/earlier `return false` case, we do not know what extension we are going to parse when more data comes in.

File: src/http/one/TeChunkedParser.cc (hunk 1 above)
Comment (rousskov) [reply src/http/one/TeChunkedParser.cc]: This concern remains unaddressed IMO, and the new `parsed the valueless chunk-ext` comment may be lying: For example, if our input contains `;name=`, then the method should return false, not true -- we have not parsed the current extension yet, and the rest of that current extension (i.e., BWS and/or the extension value) may still be coming.

File: src/http/one/Tokenizer.cc (hunk 6 above)
Comment (rousskov) [reply src/http/one/Tokenizer.cc]: The throwing condition is wrong AFAICT: false prefix() does not necessarily mean we found an invalid character. It may also mean that we found no characters at all.

File: src/http/one/Tokenizer.h (hunk 7 above)
Comment (rousskov) [reply src/http/one/Tokenizer.h]: The added documentation is lying -- a `true` return value does not necessarily mean that we have parsed a (complete or whole) token. You have added a `!tok.atEnd()` check for that case, but that check does not work the way you think it does:

```
// Distinguish complete tokens from token prefixes:
// the presence of trailing non-token characters means
// a complete token.
!tok.atEnd();
```

You are correct that the presence of a trailing non-token character means that we have parsed a complete token. However, the reverse is not true -- the lack of a trailing non-token character does not necessarily mean that we have not parsed a complete token. If the caller has received a complete input sequence, then being `atEnd()` in this context means the complete token was successfully parsed.

While a complete input sequence may not end with a token-or-quoted-string value for the current tokenOrQuotedString() caller (because chunk extensions are followed by CRLF), you are not writing this function for that specific caller. It has to work correctly for all callers. Thus, this atEof() check belongs to specific callers that can use it.

File: src/http/one/Tokenizer.h (hunk 8 above)
Comment (rousskov) [reply src/http/one/Tokenizer.h]: I am resolving this conversation. The problem with the new documentation and implementation is discussed in another (unresolved) https://github.com/squid-cache/squid/pull/310#discussion_r228577286

File: src/http/one/Tokenizer.cc (hunk 9 above)
Comment (rousskov) [reply src/http/one/Tokenizer.cc]: Not done: The following return statement (at least) modifies returnedToken when returning false: https://github.com/squid-cache/squid/blob/7c3462ebbc67447672095099c5965b67343e8ffc/src/http/one/Tokenizer.cc#L19

**Comment by rousskov** [comment]
I went through all the last review change requests and resolved those that I consider addressed. Many remain unresolved. I added a note to each of them. Please address them. I did not review the new code outside of those requests scope during this iteration. The build needs fixing as well, of course.

File: src/http/one/TeChunkedParser.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/http/one/TeChunkedParser.cc]: > This concern remains unaddressed IMO, and the new parsed the valueless chunk-ext comment may be lying: For example, if our input contains `;name=`

Disagree, the new code should work properly. Your particular `;name=` example: the execution will pass into the `if (tok.skip('=') && ParseBws(tok)) { }` block and returns false in the end.
Are there any other examples which could be misinterpreted by the new code?

File: src/http/one/Tokenizer.h (hunk 7 above)
Comment (eduard-bagdasaryan) [reply src/http/one/Tokenizer.h]: >  If the caller has received a complete input sequence, then being atEnd() in this context means the complete token was successfully parsed.

The complete input sequence may be 'complete' from the receiver of the data point of view (the sender closed the connection), but still be incomplete (e.g., the sender aborted while sending the data).

> you are not writing this function for that specific caller It has to work correctly for all callers.

I assume that the caller of this method should be specific anyway: it belongs to Http::One namespace and the callers, obviously, should belong either. Within this context (HTTP/1.1 messages), all tokens finally end up with whitespaces/line separators: if the input finished but still lacks them, the received token is invalid either.

File: src/http/one/TeChunkedParser.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http/one/TeChunkedParser.cc]: I used `the` because we are going to parse the extension belonging to the specific chunk (which size we should have parsed already). However, if you insist, I can use `an` here.

File: src/http/one/TeChunkedParser.cc (hunk 1 above)
Comment (rousskov) [reply src/http/one/TeChunkedParser.cc]: > Your particular ;name= example: the execution will pass into the if (tok.skip('=') && ParseBws(tok)) { } block and returns false in the end.

I see. ParseBws() is also broken -- it returns true when there are no characters in the input buffer. That is incorrect because it allows the code to mistakenly think that BWS was handled already (and switch to the next parsing step/phase) when, in fact, BWS will appear in the input stream after the next I/O. Please fix both bugs.

File: src/http/one/Tokenizer.h (hunk 7 above)
Comment (rousskov) [reply src/http/one/Tokenizer.h]: > The complete input sequence may be 'complete' from the receiver of the data point of view (the sender closed the connection), but still be incomplete (e.g., the sender aborted while sending the data).

Here, completeness is defined by this function caller. It does not matter to _this_ function what "complete" actually means -- it cannot (and does not want to) know how "completeness" is defined. You are right that the definition of completeness in the caller may be very complex, but that definition is caller-specific and irrelevant to this general tokenizing function.

Squid Tokenizer cannot tell whether "all data has been received" -- it simply lacks the API that can relay that information from the caller. Until that changes, Tokenizer and its knowledge-lacking users/wrappers (like this function) should not mistake an empty buffer for the promise of more token characters.

> callers, obviously, should belong either to the HTTP/1.1 messages context. Within this context, all tokens finally end up with whitespaces/line separators

No, we cannot assume that all callers are parsing an HTTP/1 message, and we cannot assume that each HTTP/1 token is followed by something else. HTTP/1 constructs such as tokens and quoted strings can be found embedded in other contexts where their length is determined by means other than whitespace and such. For a high-level example (at the whole message level), consider ICAP messages that embed HTTP headers and specify their exact size. And HTTP/2 reuses/embeds a lot of small HTTP/1 constructs, often specifying their exact size.

File: src/http/one/Parser.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/http/one/Parser.h]: My original change request described a general problem that was not exclusive to the line it was attached to. I was interpreting your "Done" in the same context and providing a link to at least one missed case.

File: src/http/one/TeChunkedParser.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/http/one/TeChunkedParser.cc]: Yes, I continue to insist. A chunk may have many extensions.

File: src/http/one/Parser.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http/one/Parser.cc]: Ok, removed Parser:: prefix from the Parser class and its children. Firstly I thought that we need only to get rid only of '::'.

File: src/http/one/Tokenizer.h (hunk 7 above)
Comment (eduard-bagdasaryan) [reply src/http/one/Tokenizer.h]: Done.

File: src/http/one/Parser.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http/one/Parser.h]: Done.

File: src/http/one/Tokenizer.cc (hunk 6 above)
Comment (eduard-bagdasaryan) [reply src/http/one/Tokenizer.cc]: Done (781dbe5).

File: src/http/one/Tokenizer.cc (hunk 9 above)
Comment (eduard-bagdasaryan) [reply src/http/one/Tokenizer.cc]: Done at 48dcb7d.

File: src/http/one/TeChunkedParser.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http/one/TeChunkedParser.cc]: Done at 781dbe5.

File: src/http/one/TeChunkedParser.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/http/one/TeChunkedParser.cc]: Done at 781dbe5.

Comment (eduard-bagdasaryan) [reply src/http/one/TeChunkedParser.cc]: > BWS will appear in the input stream after the next I/O

However, this specific case should be handled properly, I think: each chunk-ext parsing step starts with ParseBws(), thus skipping unexpected whitespaces received at each I/O.

File: src/http/one/Tokenizer.cc (hunk 6 above)
Comment (yadij) [reply src/http/one/Tokenizer.cc]: FTR because this is extremely confusing to see done with zero change. The 'done' status appears to be because the atEnd() check is done immediately after the quote is detected. So the tricky case never reaches this call to prefix().

File: src/http/one/Parser.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/http/one/Parser.cc]: Please be very careful with this removal. The ::Parser needs to be used explicitly when the generic syntax rules are being applied instead of the HTTP/1.x-specific rules (Http1::Parser). Early HTTP RFCs had some conflicting syntax requirements about tolerances being allowed or forbidden depending on the token placement in the HTTP message and sometimes even what HTTP extension header was being parsed.

Without the namespace scoping it is particularly hard to tell from any given code function which Parser is controlling the applied rules. I expect we will end up with people doing things like what you did with this PR initially: "fixing" to tolerateRFC 2616-style behaviour unknowing that the code is also handling RFC7230 coinstructs where tolerances are expressly forbidden, then changing it back, repeating indefinitely.

FWIW: HTTP/2 extends the problem further with a zero-tolerance requirement for parts of the protocol explicitly defined and inheriting BWS/RWS/OWS tolerance in other extension parts. The Http1::Parser class was existing as part of a pair with Http2::Parser class under construction to separate those distinct trees of parsing rules such that derivative Parser for other protocols could inherit the right set of rules without us needing duplicate the logic in each protocol Parser.

File: src/http/one/Tokenizer.h [hunk 14]
```diff
@@ -9,68 +9,52 @@
...
- * \see ::Parser::Tokenizer for more detail
+ *    quoted-string  = DQUOTE *( qdtext / quoted-pair ) DQUOTE
+ *    qdtext         = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
+ *    obs-text       = %x80-FF
+ *  "
+ *
+ * \param http1p0 HTTP/1.0 does not permit \-escaped characters
+ * \param tokenPrefixResult function return value when input is a token prefix
+ * \returns tokenPrefixResult if input contains nothing but a token (prefix)
+ * \returns true (and sets the value) if input starts with a token or quoted-string
```
Comment (yadij) [inline src/http/one/Tokenizer.h]: Doxygen syntax uses "retval" for lists of alternative return values. "returns" is for describing a single-type return.

File: src/http/one/Tokenizer.h [hunk 15]
```diff
@@ -9,68 +9,52 @@
...
+ * \returns false (and leaves the value intact) if input does not start with a token or quoted-string
+ * \throws on syntax violations
```
Comment (yadij) [inline src/http/one/Tokenizer.h]: Doxygen syntax requires naming the exception type which is thrown.
https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdthrow

Comment (rousskov) [reply src/http/one/Tokenizer.h]: AFAICT, Doxygen does not require naming the exception type -- doing so would actually promote bad code design. Doxygen does require naming the exception object when describing what is being thrown. In those rare cases where we should use `\throw`, starting with something like `\throws exception ...` may work well enough.

FWIW, in most cases, we should _not_ document that a function throwing at all because any function can throw by default. This function can be considered a special case, but I am not against removal of that `\throws` line if that is what other prefer.

File: src/http/one/Parser.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (eduard-bagdasaryan) [reply src/http/one/Parser.cc]: Since Http::One::Tokenizer was removed in this PR there should be no confusion: all Http::One code now uses the generic ::Parser::Tokenizer, which declared as a public typedef in Http::One::Parser.  In other words, there should be no need in duplicating "::Parser::" prefix in dozens of Http::One places, since there is a single generic Tokenizer there.

File: src/http/one/Tokenizer.h (hunk 14 above)
Comment (eduard-bagdasaryan) [reply src/http/one/Tokenizer.h]: Fixed at 3cdffa.

File: src/http/one/Tokenizer.h (hunk 15 above)
Comment (eduard-bagdasaryan) [reply src/http/one/Tokenizer.h]: To avoid confusion, I removed this `\throws` line from there (3cdffa). BTW, this is in line with other Squid code documentation, which does not use this `\throws`. I found only 1-2 uses in all Squid sources.

File: src/http/one/Tokenizer.cc (hunk 9 above)
Comment (rousskov) [reply src/http/one/Tokenizer.cc]: Still not done. For example:

https://github.com/squid-cache/squid/blob/3cdffa2b3990656edfb75142a30e7e85693192fe/src/http/one/Tokenizer.cc#L48

I know the new parseQuotedStringSuffix() function is a helper function, but I think it should follow the same "no parameter modification on false result" principle because there are no compelling reasons _not_ to follow it. The alternative is to clearly document that parseQuotedStringSuffix() is special, but I do not think introducing this general API exception is worth it.

Comment (eduard-bagdasaryan) [reply src/http/one/Tokenizer.cc]: > Still not done

Done at 3ae2217.

**Comment by eduard-bagdasaryan** [comment]
@rousskov, please review the last fixes (up to 92c894).

**Comment by rousskov** [comment]
> rousskov dismissed their [stale review](/squid-cache/squid/pull/310#pullrequestreview-194801264) [40 seconds ago](/squid-cache/squid/pull/310#event-2142939621)

Sorry, I dismissed my review in the wrong PR! You may have addressed all my change requests, but I do not know that yet. I cannot undo that action, but I will come back here shortly.

File: src/http/one/TeChunkedParser.h [hunk 16]
```diff
@@ -18,14 +18,31 @@ namespace Http
...
+
 /**
  * An incremental parser for chunked transfer coding
  * defined in RFC 7230 section 4.1.
  * http://tools.ietf.org/html/rfc7230#section-4.1
  *
  * The parser shovels content bytes from the raw
  * input buffer into the content output buffer, both caller-supplied.
- * Ignores chunk extensions except for ICAP's ieof.
+ * Ignores chunk extensions except for ICAP's use-original-body.
```
Comment (rousskov) [inline src/http/one/TeChunkedParser.h]:
```suggestion
 * Chunk extensions like use-original-body are handled via setCustomExtensionsParser().
```

File: src/http/one/TeChunkedParser.h [hunk 17]
```diff
@@ -37,6 +54,9 @@ class TeChunkedParser : public Http1::Parser
     /// set the buffer to be used to store decoded chunk data
     void setPayloadBuffer(MemBuf *parsedContent) {theOut = parsedContent;}
 
+    /// delegate parsing of some extensions to an external parser
```
Comment (rousskov) [inline src/http/one/TeChunkedParser.h]:
```suggestion
    /// instead of ignoring all chunk extension values,
    /// give the supplied parser a chance to interpret them
```

File: src/adaptation/icap/ModXact.cc [hunk 18]
```diff
@@ -1157,6 +1157,31 @@ void Adaptation::Icap::ModXact::decideOnParsingBody()
...
+
+private:
+    static const SBuf UseOriginalBodyName;
+
+    int64_t useOriginBody_;
+};
+
+const SBuf ExtensionsParser::UseOriginalBodyName("use-original-body");
+
+}
```
Comment (rousskov) [inline src/adaptation/icap/ModXact.cc]:
```suggestion
} // namespace Icap
```

File: src/adaptation/icap/ModXact.cc [hunk 19]
```diff
@@ -1157,6 +1157,31 @@ void Adaptation::Icap::ModXact::decideOnParsingBody()
...
+}
```
Comment (rousskov) [inline src/adaptation/icap/ModXact.cc]:
```suggestion
} // namespace Adaptation
```

File: src/adaptation/icap/ModXact.cc [hunk 20]
```diff
@@ -1157,6 +1157,31 @@ void Adaptation::Icap::ModXact::decideOnParsingBody()
     }
 }
 
+namespace Adaptation {
+namespace Icap {
+class ExtensionsParser : public Http1::CustomExtensionsParser
```
Comment (rousskov) [inline src/adaptation/icap/ModXact.cc]:
```suggestion
/// handles ICAP-specific chunk extensions supported by Squid
class ExtensionsParser : public Http1::CustomExtensionsParser
```

File: src/adaptation/icap/ModXact.cc [hunk 21]
```diff
@@ -1157,6 +1157,31 @@ void Adaptation::Icap::ModXact::decideOnParsingBody()
...
+        assert(knownExtension(extName));
+        return parseIntExtension(tok, UseOriginalBodyName, useOriginBody_);
+    }
+    virtual bool knownExtension(const SBuf &extName) override { return extName == UseOriginalBodyName;}
+    int64_t useOriginBody() const { return useOriginBody_; }
+
+private:
+    static const SBuf UseOriginalBodyName;
+
+    int64_t useOriginBody_;
```
Comment (rousskov) [inline src/adaptation/icap/ModXact.cc]: No need to propagate the old naming mistake:

```suggestion
    int64_t useOriginalBody_ = -1;
```

File: src/adaptation/icap/ModXact.cc [hunk 22]
```diff
@@ -1157,6 +1157,31 @@ void Adaptation::Icap::ModXact::decideOnParsingBody()
...
+namespace Adaptation {
+namespace Icap {
+class ExtensionsParser : public Http1::CustomExtensionsParser
+{
+public:
+    ExtensionsParser() : useOriginBody_(-1) {}
+
+    virtual bool parse(Tokenizer &tok, const SBuf &extName) override {
+        assert(knownExtension(extName));
+        return parseIntExtension(tok, UseOriginalBodyName, useOriginBody_);
```
Comment (rousskov) [inline src/adaptation/icap/ModXact.cc]: We should throw an error on negative values, right?

File: src/http/one/TeChunkedParser.cc [hunk 23]
```diff
@@ -115,55 +117,88 @@ Http::One::TeChunkedParser::parseChunkSize(Http1::Tokenizer &tok)
...
+
+    if (!tok.skip('='))
+        return true; // parsed a valueless chunk-ext
+
+    if (!ParseBws(tok)) {
+        tok = savedTok;
+        return false;
+    }
+
+    if (customExtensionsParser && customExtensionsParser->knownExtension(extName))
```
Comment (rousskov) [inline src/http/one/TeChunkedParser.cc]:
```suggestion
    if (customExtensionsParser && customExtensionsParser->knownExtension(extName)) {
```

Please apply automated code formatting if you have not already.

File: src/http/one/TeChunkedParser.cc [hunk 24]
```diff
@@ -13,10 +13,13 @@
 #include "http/one/Tokenizer.h"
 #include "http/ProtocolVersion.h"
 #include "MemBuf.h"
+#include "parser/Tokenizer.h"
 #include "Parsing.h"
+#include "sbuf/Stream.h"
 #include "SquidConfig.h"
 
 Http::One::TeChunkedParser::TeChunkedParser()
```
Comment (rousskov) [inline src/http/one/TeChunkedParser.cc]:
```suggestion
Http::One::TeChunkedParser::TeChunkedParser():
```

File: src/adaptation/icap/ModXact.cc [hunk 25]
```diff
@@ -1144,7 +1144,7 @@ void Adaptation::Icap::ModXact::decideOnParsingBody()
         debugs(93, 5, HERE << "expecting a body");
         state.parsing = State::psBody;
         replyHttpBodySize = 0;
-        bodyParser = new Http1::TeChunkedParser;
+        bodyParser = new Http1::TeChunkedParser();
```
Comment (rousskov) [inline src/adaptation/icap/ModXact.cc]: An unnecessary change?

File: src/adaptation/icap/ModXact.cc [hunk 26]
```diff
@@ -1157,6 +1157,31 @@ void Adaptation::Icap::ModXact::decideOnParsingBody()
...
+{
+public:
+    ExtensionsParser() : useOriginBody_(-1) {}
+
+    virtual bool parse(Tokenizer &tok, const SBuf &extName) override {
+        assert(knownExtension(extName));
+        return parseIntExtension(tok, UseOriginalBodyName, useOriginBody_);
+    }
+    virtual bool knownExtension(const SBuf &extName) override { return extName == UseOriginalBodyName;}
+    int64_t useOriginBody() const { return useOriginBody_; }
```
Comment (rousskov) [inline src/adaptation/icap/ModXact.cc]: Please split into two methods to hide our internal hack of representing abcent extension as a negative offset:

```suggestion
    bool sawUseOriginalBody() const { return useOriginalBody_ >= 0; }
    uint64_t useOriginalBody() const { assert(sawUseOriginalBody()); return static_cast<uint64_t>(useOriginalBody_); }
```

File: src/adaptation/icap/ModXact.cc [hunk 27]
```diff
@@ -1167,6 +1192,8 @@ void Adaptation::Icap::ModXact::parseBody()
     // the parser will throw on errors
     BodyPipeCheckout bpc(*adapted.body_pipe);
     bodyParser->setPayloadBuffer(&bpc.buf);
+    ExtensionsParser extensionsParser;
+    bodyParser->setCustomExtensionsParser(&extensionsParser);
```
Comment (rousskov) [inline src/adaptation/icap/ModXact.cc]: It feels wrong to (re)set the extensions parser every time we parse a piece of response body. I suggest doing it once, when we create the parser. This will require making extensionsParser a class member, but that is OK.

Making the above change will also address extensionsParser lifetime concern: This may not be possible in the current code, but, technically, since bodyParser is an incremental parser, bodyParser->parse() may return false _after_ successfully parsing and consuming the use-original-body extension value. In that case, that parsed value will be lost in the current code (because we will reset extensionsParser on the next parsing path).

**Comment by rousskov** [comment]
> I still do not like numerous repetitions of "restore tokenizer and then return false" code/logic. I still plan to suggest a solution for that separately (and have started working on that).

Done in f9876c, which simplified the code quite a bit! I have not tested my changes, but they do resolve my concerns. @eduard-bagdasaryan, please review, retest, and adjust as needed when you get a chance.

File: src/http/one/TeChunkedParser.cc [hunk 28]
```diff
@@ -104,66 +111,78 @@ Http::One::TeChunkedParser::parseChunkSize(Http1::Tokenizer &tok)
...
+}
+
+/// Parses a single chunk-ext list element:
+/// chunk-ext = *( BWS ";" BWS chunk-ext-name [ BWS "=" BWS chunk-ext-val ] )
+void
+Http::One::TeChunkedParser::parseOneChunkExtension(Tokenizer &tok)
+{
+    ParseBws(tok); // Bug 4492: ICAP servers send SP before chunk-ext-name
+
+    assert(!tok.atEnd());
```
Comment (eduard-bagdasaryan) [inline src/http/one/TeChunkedParser.cc]: I would suggest removing this assert here and in other similar contexts after ParseBws() since this method throws on atEnd() is true.

Comment (rousskov) [reply src/http/one/TeChunkedParser.cc]: Yes, please remove this assert() and any equivalent ones.

Here is some background that may help you identify equivalent asserts to remove: In earlier versions of this code, this assert was meant to clarify that the classic (non-throwing) tok.prefix() call was handled correctly: If we were to get here, and atEnd() was true before the call, then extra work would be required to distinguish insufficient input from invalid input after a false tok.prefix() call result. That earlier code did not do that extra work and attempted to clarified why not with this assert. All these complications are gone in the current code, thanks to exceptions!

**Comment by rousskov** [comment]
@yadij, do you want Factory to port these changes to v4? On one hand, the changes are substantial. On the other hand, there are a few bug fixes in this PR, including those that target bugs reachable by unlucky or specially crafted HTTP messages. Should these fixes go into v4?

**Comment by rousskov** [comment]
@yadij, do you want Factory to port these changes to v4? Please see my earlier [question](https://github.com/squid-cache/squid/pull/310#issuecomment-474473081) for details.

**Comment by yadij** [comment]
Not this late in v4 lifecycle. The size of this PR is beyond the arbitrary limits for considering backport in code size, side effects outside the tested feature, and protocol interoperability with other Squid-4. As such it is on my list of changes to hold for v5.

**Comment by rousskov** [comment]
Thank you for letting me know.
